jfr/src/main/java/io/netty/jfr/JfrChannelHandler2.java

```diff
@@ -0,0 +1,539 @@
+/*
+ * Copyright 2019 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.jfr;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandler.Sharable;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import jdk.jfr.Category;
+import jdk.jfr.Enabled;
+import jdk.jfr.Event;
+import jdk.jfr.Label;
+import jdk.jfr.Name;
+import jdk.jfr.Threshold;
+
+import java.net.SocketAddress;
+
+/**
+ * A {@link ChannelHandler} that creates JDK Flight Recorder events for all Netty channel events/operations.
+ */
+@Sharable
```

Comment:
This can not be `@Sharable` as it stores state

Reply:
The ChannelPromise stuff?

Reply:
you have the events as final fields and set the `Channel` on the events in the various callbacks. These callbacks may be executed concurrently for different `Channel`s.

Reply:
True that, will try to fix, also fixing the stacktrace if possible....

Reply:
Not sure the stacktrace can actually be fixed, since event.commit has to be run in the callback.... 
Is it worth having a different stacktrace than the non promise methods?